src/esp/gfx/PbrDrawable.cpp

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "PbrDrawable.h"
+
+#include <Corrade/Containers/ArrayViewStl.h>
+#include <Corrade/Utility/FormatStl.h>
+
+namespace Mn = Magnum;
+
+namespace esp {
+namespace gfx {
+
+PbrDrawable::PbrDrawable(scene::SceneNode& node,
+                         Mn::GL::Mesh& mesh,
+                         ShaderManager& shaderManager,
+                         const Mn::ResourceKey& lightSetupKey,
+                         const Mn::ResourceKey& materialDataKey,
+                         DrawableGroup* group)
+    : Drawable{node, mesh, group},
+      shaderManager_{shaderManager},
+      lightSetup_{shaderManager.get<LightSetup>(lightSetupKey)},
+      materialData_{
+          shaderManager.get<MaterialData, PBRMaterialData>(materialDataKey)} {
+  updateShader().updateShaderLightParameters();
+}
+
+void PbrDrawable::setLightSetup(const Mn::ResourceKey& lightSetup) {
+  lightSetup_ = shaderManager_.get<LightSetup>(lightSetup);
+  updateShader().updateShaderLightParameters();
+}
+
+void PbrDrawable::draw(const Mn::Matrix4& transformationMatrix,
+                       Mn::SceneGraph::Camera3D& camera) {
+  // no need to call updateShaderLightParameters() in the draw loop,
+  // only need to update the position or direction of the lights
+  updateShader().updateShaderLightDirectionParameters(transformationMatrix,
+                                                      camera);
+
+  (*shader_)
+      // e.g., semantic mesh has its own per vertex annotation, which has been
+      // uploaded to GPU so simply pass 0 to the uniform "objectId" in the
+      // fragment shader
+      .setObjectId(
+          static_cast<RenderCamera&>(camera).useDrawableIds()
+              ? drawableId_
+              : (materialData_->perVertexObjectId ? 0 : node_.getSemanticId()))
+      .setTransformationMatrix(transformationMatrix)  // modelview matrix
+      .setMVPMatrix(camera.projectionMatrix() * transformationMatrix)
+      .setNormalMatrix(transformationMatrix.normalMatrix())
+      .bindTextures(materialData_->baseColorTexture,
```

Comment:
`bindTextures` cannot improve the performance, but it improves the simplicity of the code here.

Reply:
I have not changed it yet. Will work on this part later.